File: YuGiOhCard/Controller/CardPicViewController.m
#import "CardPicViewController.h"
#import "CardViewController.h"
#import "CardItem.h"
#import "FileUtils.h"
#import "UrlConsts.h"

@interface CardPicViewController () {
    NSString * _img_path;
}

@property (nonatomic) CardItem * card;

@end

@implementation CardPicViewController

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    _img_path = @"image";
    self.card = ((CardViewController *)self.tabBarController).card;
    NSString * cardImgName = [NSString stringWithFormat:@"%ld.jpg", (long)self.card.card_id];
    BOOL exists = [FileUtils fileExists:cardImgName filePath:_img_path];
    if (exists) {
        [self loadImage:[FileUtils readFile:cardImgName loadPath:_img_path]];
    } else {
        [self.imgCard setHidden:YES];
        [self.btnDownload setHidden:NO];
        [self.aivDownload setHidden:YES];
        [self.btnDownload addTarget:self action:@selector(downloadClick:) forControlEvents:UIControlEventTouchDown];
    }
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
}

-(void) downloadClick:(id)sender {
    [self.aivDownload setHidden:NO];
    [self.btnDownload setHidden:YES];
    HttpUtils * hu = [HttpUtils alloc];
    hu.delegate = self;
    NSString * url = [NSString stringWithFormat:URL_CARD_IMAGE, (long)self.card.card_id];
    [hu get:url];
}

-(void) receivedData:(NSData *)data {
    if (data != nil) {
        NSString * fileName = [NSString stringWithFormat:@"%ld.jpg", (long)self.card.card_id];
        [FileUtils writeFile:fileName savePath:_img_path fileData:data];
        [self loadImage: data];
    }
}

-(void) receivedError:(NSString *)err {
    [self.imgCard setHidden:YES];
    [self.btnDownload setHidden:NO];
    [self.aivDownload setHidden:YES];
}

-(void) loadImage: (NSData *) data {
    [self.imgCard setHidden:NO];
    [self.btnDownload setHidden:YES];
    [self.aivDownload setHidden:YES];
    [self.imgCard setImage:[UIImage imageWithData:data]];
}

@end

File: swift/YuGiOhCard/Controller/CardPicViewController.swift
import UIKit

class CardPicViewController: UIViewController, HttpUtilsDelegate {

    @IBOutlet var imgCard: UIImageView?
    @IBOutlet var btnDownload: UIButton?
    @IBOutlet var aivDownload: UIActivityIndicatorView?
    
    var _img_path: String?
    var card: CardItem?
    
    var inited = false
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        if (!inited) {
            inited = true
            for temp in self.view.subviews {
                if (temp is UIImageView) || (temp is UIButton) || (temp is UIActivityIndicatorView) {
                    UIUtils.scaleComponent(view: temp)
                }
            }
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        inited = false
        _img_path = "image"
        self.card = (self.tabBarController! as! CardViewController).card
        let cardImgName = "\(self.card!._id).jpg"
        let exists = FileUtils.fileExists(fileName: cardImgName, filePath:_img_path!)
        if (exists) {
            self.loadImage(data: FileUtils.readFile(fileName: cardImgName, loadPath:_img_path!)!)
        } else {
            self.imgCard!.isHidden = true
            self.btnDownload!.isHidden = false
            self.aivDownload!.isHidden = true
            self.btnDownload!.addTarget(self, action:#selector(downloadClick(sender:)), for:UIControlEvents.touchDown)
        }
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    func downloadClick(sender: AnyObject) {
        self.aivDownload!.isHidden = false
        self.btnDownload!.isHidden = true
        let hu = HttpUtils()
        hu.delegate = self
        let url = NSString(format: URL_CARD_IMAGE, self.card!._id) as String
        hu.get(url: url)
    }
    
    func httpUtils(httpUtils: HttpUtils, receivedData data: NSData?) {
        if (data != nil) {
            let fileName = "\(self.card!._id).jpg"
            FileUtils.writeFile(fileName: fileName, savePath:_img_path!, fileData:data!)
            self.loadImage(data: data!)
        }
    }

    func httpUtils(httpUtils: HttpUtils, receivedError err: NSString) {
        self.imgCard!.isHidden = true
        self.btnDownload!.isHidden = false
        self.aivDownload!.isHidden = true
    }
    
    func loadImage(data: NSData) {
        self.imgCard!.isHidden = false
        self.btnDownload!.isHidden = true
        self.aivDownload!.isHidden = true
        self.imgCard!.image = UIImage(data: data as Data)
    }
}
